### PR TITLE
Track mixpanel events for billing and subscription actions

### PR DESCRIPTION
**Pull request by Devorein**
### <samp>🤖 Generated by Copilot at 04a8710</samp>

This pull request implements various improvements and updates to the subscription feature of the app. It enhances the checkout form, the subscription information, and the subscription settings components, and tracks more user events with Mixpanel. It also handles and tracks Stripe webhook events more accurately, and updates the types for the subscription events. It also reorders some imports in a minor change.

### CODE REVIEW

File: lib/metrics/mixpanel/interfaces/SubscriptionEvent.ts [hunk 1]
```diff
@@ -1,21 +1,47 @@
...
-
 import type { SubscriptionPeriod } from 'lib/subscription/constants';
+import type { PaymentMethodType } from 'lib/subscription/mapStripeFields';
 
 import type { BaseEvent } from './BaseEvent';
 
-export type ViewSubscriptionEvent = BaseEvent;
-export type InitiateSubscriptionEvent = BaseEvent;
+export type CreateSubscriptionEvent = BaseEvent & {
+  paymentMethod: PaymentMethodType;
```
Comment (motechFR) [inline lib/metrics/mixpanel/interfaces/SubscriptionEvent.ts:7]: I think we should add subscription ID here

File: pages/api/v1/webhooks/stripe/index.ts [hunk 2]
```diff
@@ -141,21 +144,30 @@ export async function stripePayment(req: NextApiRequest, res: NextApiResponse):
...
-        }
 
-        log.info(
-          `The invoice number ${invoice.id} for the subscription ${stripeSubscription.id} was paid for the spaceId ${spaceId}`
-        );
+          trackUserAction('create_subscription', {
+            blockQuota,
+            period,
+            spaceId,
+            paymentMethod: 'card',
```
Comment (motechFR) [inline pages/api/v1/webhooks/stripe/index.ts:152]: We probably should update this to account for different payment methods, and not hardcode it

**Review by motechFR — APPROVED** [review]

File: components/settings/subscription/CheckoutForm.tsx [hunk 3]
```diff
@@ -63,12 +64,19 @@ export function CheckoutForm({
     resolver: yupResolver(schema())
   });
 
+  useEffect(() => {
+    if (space) {
+      charmClient.track.trackAction('view_checkout_screen', {
```
Comment (mattcasey) [inline components/settings/subscription/CheckoutForm.tsx:69]: I think this should still be page_view, with a specific type for checkout

**Review by mattcasey — CHANGES_REQUESTED** [review]
We should stick with 'page_view' when tracking views. The type could be 'billing/marketing', 'billing/checkout', 'billing/settings', etc

File: lib/metrics/mixpanel/interfaces/SubscriptionEvent.ts (hunk 1 above)
Comment (Devorein) [reply lib/metrics/mixpanel/interfaces/SubscriptionEvent.ts:7]: Nice one. I will add it

File: pages/api/v1/webhooks/stripe/index.ts (hunk 2 above)
Comment (Devorein) [reply pages/api/v1/webhooks/stripe/index.ts:152]: Correct. I have contacted valentin and we will get this fixed.

File: components/settings/subscription/CheckoutForm.tsx (hunk 3 above)
Comment (Devorein) [reply components/settings/subscription/CheckoutForm.tsx:69]: Lets go with that then

**Review by mattcasey — APPROVED** [review]

File: pages/api/v1/webhooks/stripe/index.ts [hunk 4]
```diff
@@ -225,6 +247,19 @@ export async function stripePayment(req: NextApiRequest, res: NextApiResponse):
...
+          const activeSubscription = await getActiveSpaceSubscription({ spaceId });
+          if (activeSubscription) {
+            trackUserAction('update_subscription', {
+              blockQuota: subscription.items.data[0].quantity as number,
+              period: subscription.items.data[0].price.recurring?.interval === 'month' ? 'monthly' : 'annual',
+              previousBlockQuota: activeSubscription.blockQuota,
+              previousPeriod: activeSubscription.period,
+              subscriptionId: activeSubscription.id,
+              spaceId,
+              userId: ''
```
Comment (mattcasey) [inline pages/api/v1/webhooks/stripe/index.ts:260]: why do we ened to pass in an empty string here? are the types wrong?

File: pages/api/v1/webhooks/stripe/index.ts [hunk 5]
```diff
@@ -301,6 +343,52 @@ export async function stripePayment(req: NextApiRequest, res: NextApiResponse):
...
+              ? 'crypto'
+              : charge.payment_method_details?.type?.startsWith('ach')
+              ? 'ach'
+              : 'card',
+            userId: ''
+          });
+        }
+
+        const lastFinalizationError = invoice.last_finalization_error;
+        log.error(`Invoice payment failed for invoice ${invoice.id} for the subscription ${stripeSubscription.id}`, {
```
Comment (mattcasey) [inline pages/api/v1/webhooks/stripe/index.ts:376]: this should be a warning, since it's not an error from our application

Comment (Devorein) [reply pages/api/v1/webhooks/stripe/index.ts:376]: Thank you. Updated

File: pages/api/v1/webhooks/stripe/index.ts (hunk 4 above)
Comment (Devorein) [reply pages/api/v1/webhooks/stripe/index.ts:260]: Right. userId is a required field and used as the distinct_id when creating the mixpanel event. I will just go with space.createdBy for now

**Pull request merged**